### PR TITLE
Fix stacking of dashcards

### LINKED ISSUE #31274: DashCardActionsPanel is clipping under other dash-cards
### Describe the bug

The `DashCardActionsPanel` for adding series to dash-cards, editing dash-card visualization settings, previewing/editing text cards, deleting cards, etc. clips under other dash-cards.
image
[](url)

### To Reproduce

1. Go to any dashboard populated with dash cards
2. Edit the dashboard
3. Hover over any dash card that has a dash card above it
4. You should see the dash card actions panel clipping under the dash card above it
  a. If you do not, move the top dash card below the bottom dash card and hover over the bottom dash card

dash-actions-panel-clipping

### Expected behavior

The dash card actions panel should always be completely visible over other dash cards

### Logs

_No response_

### Information about your Metabase installation

```JSON
- Version: master
- Browser: Chrome
```

### Severity

P3: Non-blocking cosmetic bug

### Additional context

_No response_

### LINKED ISSUE #32827: Dashboard visualization buttons not visible when focused with keyboard
### Describe the bug

image

### To Reproduce

https://github.com/metabase/metabase/assets/6830683/ca904ee5-deb6-453c-bf9e-5e9346f324f2

1. Create a dashboard
2. Edit it
3. Add 2 text blocks
4. Put the 2nd one under the 1st one
5. Focus the 1st one
6. Hit <kbd>Tab</kbd>

Focused button is not visible but a tooltip is shown for it.

### Expected behavior

Focused button is visible and a tooltip is shown for it

### Severity

P4

### Additional context

Found when reviewing #32796

### PR DESCRIPTION
**Pull request by uladzimirdev**
Closes https://github.com/metabase/metabase/issues/31274, resolves #32827

### Description

This change actually fixes stacking cards on a dashboard. 
Includes refactoring of the dashcard css: some styles were moved from .css to emotion

### How to verify

1. Go to dashboard edit mode or create a new one, add three cards on it, change their order, hover over card to see the action buttons -> make sure they are not cut

Added e2e test which relies on the ability of cypress to click on not covered elements only.

### Demo

https://www.loom.com/share/d8306edc254246cdb69a3bb874c9b1f5

### Checklist

- [ ] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Comment by uladzimirdev** [comment]
@WiNloSt do you see any problem such fix? do you have an idea about the possible tests for this issue?

File: frontend/src/metabase/css/dashboard.css [hunk 1]
```diff
@@ -101,6 +101,14 @@
   z-index: 2;
 }
 
+/* dashcards use positioned absolutely,
+   so children calculate z-index inside a container, not globally,
+   that's why we have to override parent's z-index
+*/
+.DashCard:hover {
+  z-index: 3;
+}
```
Comment (WiNloSt) [inline frontend/src/metabase/css/dashboard.css:110]: Clever, I like the idea. This is solution is really simple and I like it. However, let's not add anything to CSS files as they're deprecated.

I think the class name is added here
https://github.com/metabase/metabase/blob/8e1f884aa2d1a785efb22a46e1dca8cb0e3fb56c/frontend/src/metabase/dashboard/components/DashboardGrid.jsx#L401-L407

`<DashboardCard />` is already an emotion component, you should be able to put this CSS rule there instead.

As for other works regarding CSS files, I've seen.

Splitting CSS files or removing rules are fine, but adding one will be questionable.

**Review by WiNloSt — COMMENTED** [review]
The solution looks great, but please move the CSS rule to emotion.

Re: testing can you try that without the `data-testid="dashboardcard-actions-panel"` will not be visible. i.e. `cy.findByTestId("dashboardcard-actions-panel").should("be.visible")` will fail if we hover over the dash card. I know Cypress try to be smart and see if the element under assertion is covered by any other elements or not. But I'm not sure how it deals with such a situation when the element is only covered partially.

If that doesn't work I'll ask Nemanja for advice. We have had tests that test CSS property in the past which I don't quite like but it's a compromise that even we can't assert the actual behavior, we can still be partially confidence that the if the element has some CSS properties applied, it should do it's job.

File: frontend/src/metabase/css/dashboard.css [hunk 2]
```diff
@@ -101,6 +101,14 @@
   z-index: 2;
 }
 
+/* dashcards use positioned absolutely,
+   so children calculate z-index inside a container, not globally,
+   that's why we have to override parent's z-index
+*/
```
Comment (kamilmielnik) [inline frontend/src/metabase/css/dashboard.css:107]: I feel like this comment makes sense only given the full context of the PR and knowing previous attempts to solve the problem. The fix is straightforward, I don't need this comment to understand it.

> so children calculate z-index inside a container, not globally,

A nicer way to put this: "so it forms a new [stacking context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_positioned_layout/Understanding_z-index/Stacking_context)".



```suggestion
/**
 * Dashcards are positioned absolutely so each one forms a new stacking context.
 * The dashcard user is currently interacting with needs to be positioned above other dashcards
 * to make sure it's not covered by absolutely positioned children of neighboring dashcards.
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_positioned_layout/Understanding_z-index/Stacking_context
 */
```

**Review by kamilmielnik — COMMENTED** [review]
I was trying to think how we can test this. Would it be possible to click particular coordinates (the area of the button that gets covered underneath a dashcard)?

File: frontend/src/metabase/css/dashboard.css (hunk 2 above)
Comment (uladzimirdev) [reply frontend/src/metabase/css/dashboard.css:107]: great improvement, thanks!

File: frontend/src/metabase/css/dashboard.css (hunk 1 above)
Comment (uladzimirdev) [reply frontend/src/metabase/css/dashboard.css:113]: I kinda agree with you that we shouldn't use css files at all, at the same time without the one large refactoring of dashcards css I doubt splitting it between css file and styled file will add any benefit. Let me check how complicated removal of css file will be

Comment (WiNloSt) [reply frontend/src/metabase/css/dashboard.css:113]: Fair, if it's not straightforward. I think we can justify leaving it here in this case.

File: e2e/test/scenarios/dashboard/reproductions/31274-dashcard-actions-overlap.cy.spec.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+describe("issue 31274", () => {
+  beforeEach(() => {
+    restore();
+    cy.signInAsAdmin();
+  });
+
+  // cypress automatically scrolls to the element, but we don't need it in this test
+  it(
+    "should not clip dashcard actions (mestabase#31274)",
+    { scrollBehavior: false },
```
Comment (uladzimirdev) [inline e2e/test/scenarios/dashboard/reproductions/31274-dashcard-actions-overlap.cy.spec.js:44]: @nemanjaglumac what do you think about such approach of avoiding `scroll` in the view? I'd like to see all dashcards in the viewport, but if I just use realHover or trigger mouseover, cypress scrolls and hides first dashcard

File: frontend/src/metabase/css/dashboard.css (hunk 1 above)
Comment (uladzimirdev) [reply frontend/src/metabase/css/dashboard.css]: I moves some styles, which was easy to test

File: e2e/test/scenarios/dashboard/reproductions/31274-dashcard-actions-overlap.cy.spec.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+        visitDashboard(dashboard.id);
+        editDashboard(dashboard.id);
+
+        secondTextCard().realHover();
+
+        visibleActionsPanel().should("have.length", 1);
+
+        cy.log(
+          "Make sure cypress can click, so element is not covered by another",
```
Comment (kamilmielnik) [inline e2e/test/scenarios/dashboard/reproductions/31274-dashcard-actions-overlap.cy.spec.js:60]:
```suggestion
          "Make sure cypress can click the element, which means it is not covered by another",
```

**Review by kamilmielnik — APPROVED** [review]
LGTM :+1:

File: frontend/src/metabase/dashboard/components/DashboardGrid.styled.tsx [hunk 5]
```diff
@@ -6,10 +6,55 @@ interface DashboardCardProps {
...
+  z-index: 2;
+
+  /**
+  * Dashcards are positioned absolutely so each one forms a new stacking context.
+  * The dashcard user is currently interacting with needs to be positioned above other dashcards
+  * to make sure it's not covered by absolutely positioned children of neighboring dashcards.
+  *
+  * @see https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_positioned_layout/Understanding_z-index/Stacking_context
+  */
+  &:hover {
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/components/DashboardGrid.styled.tsx:19]: Should we add `:focus-within` here too? So when we navigate to the dashcard actions using keyboard it won't be covered by the card above.

It'd be great to add tests to cover this case too.

```suggestion
  &:hover, &:focus-within {
```

#### This PR
Screenshot 2023-08-08 at 3 26 57 PM

#### With `:focus-within`
Screenshot 2023-08-08 at 3 27 19 PM

**Review by WiNloSt — COMMENTED** [review]
I found a regression in link cards. The dashcard action is covered by the text input. I'm not sure if this is caused by this PR, but could you take a look?

#### This PR
Screenshot 2023-08-08 at 3 25 46 PM


#### Master
Screenshot 2023-08-08 at 3 26 23 PM

**Comment by uladzimirdev** [comment]
Sure, I’ll take a look

Вт, 8 авг. 2023 г. в 11:45, Mahatthana (Kelvin) Nomsawadi <
***@***.***>:

**Pull request merged**